Bigboxscraper+rdescraper to selenium

JuozapasG opened the pull request:
padaryti bigbox ir rde scraper
bsk patvarkiau koda keliose mini vietose

JuozapasG committed:
bigbox scraper to selenium
little code fix

JuozapasG committed:
perrasytas rde scraperis ant selenium
istaisiau ermitazasscaper -> ermitazasscraper

JuozapasG committed:
working scrapers

JuozapasG committed:
.

DuckDrick commented on Comparison shopping engine/Forms/SearchForm.cs:
```diff
@@ -100,9 +100,7 @@ private void ToggleSourcePanel(object sender, EventArgs e)
 
         private void SearchForm_Load(object sender, EventArgs e)
```
Jeigu metodas tuščias tai ar nebūtų galima jo visai ištrinti?

JuozapasG replied:
cia form load metodas, be jo ismes errora

DuckDrick replied:
Tai kaip aš suprantu errora meta dėl to, kad šis metodas yra priskirtas tai formai, jeigu nebūtų tai ir nemestų

DuckDrick commented on Comparison shopping engine/Forms/SearchForm.cs:
```diff
@@ -154,15 +152,18 @@ private int CountHowMany<T>(T value)
+            return _items.Count(product => product.Source.Contains(value.ToString()));
 
         }
         private void SearchForm_Shown(object sender, EventArgs e)
```
Gal vietoj to, kad viską užkraudinėtų metode kuomet langas jau pasijungia reiktų visą tai daryti kuomet langas kraunasi?
Turiu omenyje perkelti iš SearcForm_Shown metodo į load metodą

JuozapasG replied:
Gera įdėja, tiesiog perkelsiu metodus.

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/AbstractSeleniumScraper.cs:
```diff
@@ -84,7 +84,7 @@ public void ScrapeWithSelenium(object s)
                                                     photoUrl,
                                                     "None",
-                                                    site + ".lt"));
+                                                    site)); // + ".lt"
```
Kokiu tikslu nuimtas nuo produkto šaltinio puslapio .lt galas?

JuozapasG replied:
Darant bigbox selenium scraperi, atsirado problema, jog pridedinėjo itemus kaip bigbox.lt, o ne bigbox.
Kadangi visi kiti saltiniai buvo be .lt galo tai nuėmiau si gala, is viskas veikia, nes nedėjo pirma i database bigbox produktu. Finalas toks jog palikau uzkomentaves, nes gali atsirasti koks bugas su kitais scraperiais, bet kiek testavau problemų nebuvo. Kiek pamenu sitas .lt galas buvo reikalingas, kuomet dar senoje mainformoje dėjome produktus i listview ir tenais parasydavom is kokio tinklapio paimta su .lt, bet dabar to nebedarom.

DuckDrick replied:
Nežinau kaip tai galėjo turėti įtakos dėjimui i db
![image](https://user-images.githubusercontent.com/67711718/97223751-de9ae500-17d8-11eb-984e-fddfe6c485eb.png)

Kaip nuotraukoje matosi, dedant i db nurodomas "site" kas yra tik tinklapio pavadinimas, be jokių .lt ar www., mano komentuotam kode aš jeigu gerai pamenu kaip veikia, tai ten tiesiog dėl estetikos, kad pridejus i listview'a rasytų *.lt

JuozapasG replied:
Na galbūt kažką sumaišiau, bet paliksiu taip, nes is db pridedami duomenys yra be .lt

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/AutoaibeScraper.cs:
```diff
@@ -79,8 +79,7 @@ protected override bool ShouldScrapeIf(IWebElement product)
 
         protected override (string, string, string, string) GetInfo(IWebElement product)
```
Ar nereiktų viską sudėti į try catch blokus, kad netyčia atsitikus tokiam atvėjui, jog nerado to produkto informacijos nemestų exceptiono? Ar kažkaip kitaip su tuo susidoroti.

JuozapasG replied:
galbūt būtų paprasciau abstract klaseje pridėti dar viena if'a po sito metodo iskvietimo ir patikrinti ar kazkuris laukas yra null, jeigu taip, tuomet ta produkta praleistume, ka manai?Tuomet nereikės eiti per visa scraperiu klases kad pridėti try catch. Lauksiu atsakymo ir jeigu tinka pridėsiu tikrinimą. Arba galima būtų ir abstract klasej dėti trycatch, bet nezinau ar taip suveiktu. 
🤔

DuckDrick replied:
Manau pavyktų su try catch'u abstrakčioje klasėje

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/BigboxScraper.cs:
```diff
@@ -0,0 +1,83 @@
+    {
+        protected override void NavigateToNextPage(ChromeDriver driver)
+        {
+            var next = driver.FindElements(By.CssSelector("li#pagination_next_bottom.pagination_next"));
```
Vietoj viso šio css selectoriaus nebūtų galima ieškoti tiesiog pagal mygtuko id?

JuozapasG replied:
Jaučiuosi saugiau dirbdamas su css selectoriumi, isbandyta, veikia :D

DuckDrick replied:
Truputėlį blogai supratai matyt, klausiau ar neužtektų parašyti "#pagination_next_bottom", nes jeigu neklystu html dokumente tik vienas elementas gali turėti tam tikrą id

JuozapasG replied:
Aaa, tuomet gal ir užtektu, bet čia dėl atsargos nurodžiau tikslesnį path.

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/BigboxScraper.cs:
```diff
@@ -0,0 +1,83 @@
+            {
+                foreach (var span in spanList)
+                {
+                    if (!span.Text.Equals("Pradžia") && !span.Text.Equals("BIGBOX"))
```
Nebūtų įmanoma tiesiog pasiimti 3 elementą iš masyvo (Jeigu jis egzistuoja žinoma)?

JuozapasG replied:
Thats smart! Taisau.

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/BigboxScraper.cs:
```diff
@@ -0,0 +1,83 @@
+        {
+            if (product.FindElements(
+                    By.CssSelector(
+                        "span.category-item-buttons.button.ajax_add_to_cart_button.disabled.btn.btn_default"))
```
Vėlgi, ar tikrai reikalingas toks ilgas selectorius?

JuozapasG replied:
na kiek testavau, toks veikia, jį sumazinus prideda produktus, kurie yra išparduoti, galbūt ir būtų galima kazkuria būtent dalį nukirpti, tačiau manau geriau nerizikuot.
edit  - testuoju vėl, neveikia... Ieškau sprendimo.

JuozapasG replied:
fixed, neišėjo labai sutrumpinti, bet veikia viskas, kazkaip įdomiai čia su bigboxo mygtukais

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/BigboxScraper.cs:
```diff
@@ -0,0 +1,83 @@
+
+        protected override (string, string, string, string) GetInfo(IWebElement product)
+        {
+            var price = product.FindElement(By.CssSelector("span.price.product-price")).Text;
```
Vėl tas pats reikalas dėl neradimo ir try catch'o

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/RdeScraper.cs:
```diff
@@ -0,0 +1,83 @@
+        {
+            string beforeUrl = driver.Url;
+            string[] currentUrl = driver.Url.Split('/');
+            var url = "https://rde.lt/search/lt/word/" + currentUrl[6] + "/page/" +
```
var nuorodos = driver.FindElements(By.CssSelector(".pages"));
if (nuorodos.Count > 0){
var puslapiai = nuorodos.FindElements(By.TagName("a"));
}
var nextPage = puslapiai[puslapiai.Count-2];

Ar kartais netiktų toks kodas?

JuozapasG replied:
Kodėl būtent .pages ? Nelabai supratau sitos dalies
Mano logika tuomet rastu tik 2 .pages, virsutine ir apatine pasirinkimų juostą, tai reiskia ju countas niekad nebus didesnis už 3. Iš to seką jog nepateksim i kitą puslapį. Nebent iškarto ieškotume su tagname "a", bet tada reiktu daryti kitokį if'ą manyciau.
Pagrinde kodėl rasiau toki ilgą koda, tai nes kito puslapio pasirinkimas, neturi aiskaus tagname ar class. O nuėjus iki paskutinio puslapio išviso dingsta tas pasirinkimas. Šiuo atveju manau ir "var nextPage = puslapiai[puslapiai.Count-2];" netiktu.
Gal galėtum labiau paaiškinti, jeigu kažko nesupratau?

DuckDrick replied:
oj, pats suklydau, 
var puslapiai = nuorodos[0].FindElements(By.TagName("a"));
turėjo atrodyti sitaip
bet vistiek dar kartelį patikrinau, toks kodas neveiktų, nes paskutiniuose puslapiuose tos rodyklėlės nebėra kuri permestų į sekantį puslapį

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/TopocentrasScraper.cs:
```diff
@@ -69,8 +69,7 @@ protected override bool ShouldScrapeIf(IWebElement product)
+            var price = product.FindElement(By.ClassName("Price-price-27p")).Text;
             var name = product.FindElement(By.ClassName("ProductGrid-productName-1JN")).Text;
             var productUrl = product.FindElement(By.ClassName("ProductGrid-link-3Q6")).GetAttribute("href");
             string photoUrl ="";
```
Kitur pakeitei į var, o čia palikai 😢

JuozapasG replied:
Nenorėjau, jaučiuosi blogai dėl photoUrl... 😿

DuckDrick commented on Comparison shopping engine/Scrapers/Selenium/VarleScraper.cs:
```diff
@@ -70,15 +70,10 @@ protected override bool ShouldScrapeIf(IWebElement product)
-            //{
-            //    photoUrl = "https://upload.wikimedia.org/wikipedia/commons/0/0a/No-image-available.png";
-            //}
+            string photoUrl = " ";
```
Kaip ir čia 😢

DuckDrick commented on Comparison shopping engine/Values/Values.cs:
```diff
@@ -5,8 +5,8 @@ namespace Comparison_shopping_engine
-            /*rde,          */ "none",
-            /*bigbox,       */ "none",
+            /*rde,          */ "https://rde.lt/search_result/lt/word/☼/page/1",
+            /*bigbox,       */ "https://bigbox.lt/paieska?controller=search&orderby=position&orderway=desc&ssa_submit=&search_query=",
```
Neužtektų "https://bigbox.lt/paieska?search_query=" ?

JuozapasG replied:
Geras pastebėjimas, dėkui kad pabandei!

DuckDrick replied:
👍

DuckDrick reviewed (commented):
Kilo keletas klausimą, porą iš jų pasikartojančiu, dėl to kode vienoduose vietose kitur gali būti ir nepakomentuota 🤔

JuozapasG committed:
some fixes on different things that were commented

JuozapasG committed:
bigbox should scrape and photo fix

JuozapasG committed:
tried to make loading while scraping
didnt succed

JuozapasG committed:
Sukurtos labiau išsamios klasės, skirtos filtrauoti produktams.
Filtravimas veikia
Reikia pridėti jog rodytu kiek yra produktų is kiekvienos kategorijos.

JuozapasG committed:
Sutvarkytas filtravimas paieškos formoje, dabar rodo kiek yra prekių skirtingame grupiu filtre.

JuozapasG committed:
updates about and faq

JuozapasG committed:
.
